skbio/metadata/_interval.py

```diff
@@ -0,0 +1,205 @@
+        old_feature : Feature
+            Original feature to search and replace in the IntervalMetadata
+        new_feature : Feature
+            The new feature to replace the original feature.
+        """
+        ivs = self.features[old_feature]
+        self.features[new_feature] = self.features.pop(old_feature)
+        for iv in ivs:
+            start, end = _polish_interval(iv)
+            self.intervals.update(start, end, old_feature, new_feature)
+
+    def add(self, feature, *intervals):
+        """ Adds a feature to the metadata object.
+
+        Parameters
+        ----------
+        feature : skbio.sequence.feature
+            The feature object being added.
+        intervals : iterable of intervals
+            A list of intervals associated with the feature
+
+        """
+        if not isinstance(feature, Feature):
+            raise ValueError('feature is not an instance of `Feature`')
+
+        # TODO: The below will require careful consideration
+        # since this will be using a little more memory, since the whole
+        # feature object will be stored.
+        for interval in intervals:
+            loc = _polish_interval(interval)
+            if loc is not None:
+                start, end = loc
+                self.intervals.add(start, end, feature)
+
+        # TODO: The below will require careful consideration
+        # since this will be using a little more memory, since the whole
+        # feature object will be stored.
+        if feature not in self.features.keys():
+            self.features[feature] = []
+
+        self.features[feature] = list(map(_polish_interval, intervals))
+
+    def _query_interval(self, interval):
+        start, end = _polish_interval(interval)
+        features = self.intervals.find(start, end)
+        return features
+
+    def _query_feature(self, key, value):
+        queries = []
+        for feature in self.features.keys():
+            if feature[key] == value:
+                 queries.append(feature)
+        return queries
+
+    def query(self, *args, **kwargs):
+        """ Looks up features that with intervals and keywords.
+
+        Parameters
+        ----------
+        args : I1, I2, ...
+            Iterable of tuples or Intervals
+        kwargs : dict
+            Keyword arguments of feature name and feature value, which can
+            be passed to ``dict``.  This is used to specify the search parameters.
+            If the `location` keyword is passed, then an interval lookup will be
+            performed.
+
+        Note
+        ----
+        There are two types of queries to perform
+        1. Query by interval
+        2. Query by key/val pair (i.e. gene=sagA)
+
+        Returns
+        -------
+        list, Feature
+            A list of features satisfying the search criteria.
+        """
+        feats = []
+
+        # Find queries by interval
+        for value in args:
+            feats += self._query_interval(value)
```

Comment:
why not have `feats = set()`, and add to it with `feats.update(...)`? 

the return is currently a list, but that suggests order matters. if order does matter, then the current approach will disregard order due to the cast to `set` on return